### DISCUSSION TITLE
Do validators have to return?

### DISCUSSION THREAD

**Discussion by FerusAndBeyond** [pydantic/pydantic#8493] [Feature Request]
`AfterValidate`, etc. has to return the value, even if it's not transformed. Is there anything that *just* checks if a condition is fulfilled?

Something like:

```
from typing import Annotated
from pydantic.functional_validators import AfterValidator
from pydantic import BaseModel

def Verify(assertion_method, msg=""):
    def wrapper(v):
        if not assertion_method(v):
            raise ValueError(msg)
        return v
    return AfterValidator(wrapper)

class Test(BaseModel):
    some_string: Annotated[str, Verify(lambda x: "hi" in x, "Must contain 'hi'")]

print(Test(some_string="hi"))
print(Test(some_string="hello"))
# => Value error, Must contain 'hi'
```

**Comment by RS-Credentive** [comment]
Are you just trying to write an error to console/log? What do you want the function to do if an error is detected? You can add a validator that takes self and returns self. I use that construct to warn people using deprecated fields or values in a model.

**Reply by RS-Credentive** [reply]
Note this will only work with an "after" validator

**Reply by RS-Credentive** [reply]
Example of a warning for a deprecated field value:
```
    @field_validator("depends_on", mode="after")
    @classmethod
    def depends_on_deprecated(
        cls, depends_on: str
    ) -> str
        # raise a DeprecationWarning if depends-on is present
        warnings.warn(
            "depends-on is a deprecated field for catalog", DeprecationWarning
        )
        return depends_on
```

**Reply by FerusAndBeyond** [reply]
For my use case, it's an endpoint in FastAPI. Thus, when data is received from the client it needs to check if it follows the requirements, e.g. username is not too long, otherwise it should throw bad request.

**Reply by RS-Credentive** [reply]
Oh, I don't know FastAPI. The validator could raise an Exception instead of returning a value. Can you raise an Exception and catch it in an upstream function somewhere in FastAPI? If a code path raises an exception, it doesn't have to return a value.
